### PR TITLE
Add type definitions for apca-w3 module

### PR DESCRIPTION
**Pull request by daneden**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by daneden** [comment]
Unfortunately this is failing because of the fact that the apca-w3 module itself is not versioned using semver 🥲 would be great to get some help with this CI failure!

**Comment by Myndex** [comment]
> _Unfortunately this is failing because of the fact that the apca-w3 module itself is not versioned using semver 🥲 would be great to get some help with this CI failure!_

Dang sorry, I had already planned on fixing the version number the next time I publish the package, and getting rid of the internal build codes, sorry that it caused a problem. The next will be 0.1.0, trying to get it out today.

**` colorparsley `** which is stable as its own package now, is **going to be removed from the APCA JS script**, and listed in the JSON package file as a dependency — colorparsley is optional for APCA. The key critical functions for APCA ae sRGBtoY, dP3toY and adobe98toY so are all part of the APCA method and JS file. 

The "...toY()" functions will all take a simple three element array of positive 8 bit numbers [255,255,255] and will output a positive number 0.0 to 1.1 for **`Ys`** (will be higher more when HDR becomes supported). (note that these functions do nothing with the alpha — alpha processing will be somewhere else).

The APCAcontrast() takes text `Ys`, background `Ys`, and an optional `places` which determines output type. The default -1 returns a signed number ± 108.0 ~ish for the output Lc values (lightness contrast)  - sending the `places` parameter an int of 0 or more and APCA returns a string of the  rounded Lc value.

I don't anticipate any of these in/out types to change for the **apca-w3** package at this point.

I'm trying to get the updated APCA package published today or tomorrow.

Thank you!

**Comment by Myndex** [comment]
The npm for apca-w3 0.1.0 is published — sorry for the issue!!

File: types/apca-w3/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * Convert a color to its rgba components
+ *
+ * @param colorIn
+ * The color to convert. Can be a string (e.g. "#ffe4c4", "aliceblue"),
+ * hexadecimal integer (0xFF0000), rgba array, or an Object with keys `r, g, b, a`
+ * and number values.
+ *
+ * @returns
+ * rgba components in a float array (`[r: number, g: number, b: number, a: number])
+ */
```
Comment (peterblazejewicz) [inline types/apca-w3/index.d.ts:54]:
```suggestion
```
seems like omission, module exports 3 functions

File: types/apca-w3/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ *
+ * @param txtY
+ * The linear Y (luminance) of the text color. Must be between 0.0-1.1.
+ *
+ * @param bgY
+ * The linear Y (luminance) of the background color. Must be between 0.0-1.1.
+ *
+ * @param places
```
Comment (peterblazejewicz) [inline types/apca-w3/index.d.ts:14]:
```suggestion
 * @param [places]
```

File: types/apca-w3/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ * of a color
+ *
+ * @returns
+ * The linear Y (luminance) of the given rgba color (+0.0-1.1)
+ */
+export function sRGBtoY(rgba: number[]): number;
+
+/**
+ *
+ * @param rgba
```
Comment (peterblazejewicz) [inline types/apca-w3/index.d.ts:35]:
```suggestion
 * @param [rgba]
```

File: types/apca-w3/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ * An array of numbers representing the red, green, blue, and alpha components
+ * of a color in the p3 display color space
+ *
+ * @returns
+ * The linear Y (luminance) of the given rgba color (+0.0-1.1)
+ */
+export function displayP3toY(rgba: number[]): number;
```
Comment (peterblazejewicz) [inline types/apca-w3/index.d.ts:42]:
```suggestion
export function displayP3toY(rgba?: number[]): number;
```
the `rgba` seems to be optional, defaulting to [0,0,0,1.0]
btw: you can type them as tuple `[number, number, number, number]` or named tuples `[r: number, g: number, b: number, a: number]`
https://www.typescriptlang.org/docs/handbook/2/objects.html#tuple-types

**Review by daneden — COMMENTED** [review]
Thanks for the review @peterblazejewicz! Merged your suggestions.

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@daneden thx!

**Comment by Myndex** [comment]
Hi @peterblazejewicz  and @daneden 

Thank you for all of this, and sorry for any growing pains as this project moves from an alpha research project to this public beta.

> _the rgba seems to be optional, defaulting to [0,0,0,1.0]
btw: you can type them as tuple [number, number, number, number] or named tuples [r: number, g: number, b: number, a: number]_

So, ` sRGBtoY() `, ` displayP3toY() `, and ` adobeRGBtoY() `   all take a simple array of which the first three elements are numbers (as of right now, ints 0-255, except displayP3toY() which takes floats 0.0-1.0). The elements are accessed by indexes 0 1 2

The idea is the simplest that can be iterated... **so I have a question,** in reading the docs it appears that with named tuples the numbers to iterate would be 1 3 5 ?

Some other notes:
- The arrays passed to these functions may have additional elements all of which are ignored - only elements 0 1 2 are used.
- The " ...toY" functions all ignore any alpha, as alpha processing is separate in ` alphaBlend() ` where the first array is rgba and the second is rgb only.
- ` calcAPCA() ` requires colorparsley, and the first two params are textColor (optional alpha) and bgColor (ignores alpha), either of which can be a string, a number (ox123456 or 8765432), a simple array of numbers, or an object with named elements (in the form of either r: or red:)

## _Question:_
What can I do to be helpful here? Are there some .ts or other files I should include in each package to make this process easier here?


Thank you!

Andy

**Comment by daneden** [comment]
> The arrays passed to these functions may have additional elements all of which are ignored - only elements 0 1 2 are used.

I actually avoided changing the type signature for the `[…]toY()` functions from `number[]` to tuples because of this!

> The " ...toY" functions all ignore any alpha, as alpha processing is separate in alphaBlend() where the first array is rgba and the second is rgb only.

I didn’t realise this, I can update the signature/docs to reflect this nuance

> calcAPCA() requires colorparsley […]

I didn’t include this function in the type definitions but can add it in the next revision. I think since colorParsley is used in the implementation but not in the type definition it shouldn’t be a problem, but we can document the requirement in the type documentation.

> […] in reading the docs it appears that with named tuples the numbers to iterate would be 1 3 5 ?

@Myndex named tuples are still just comma-separated arrays, so these are equivalent:

```typescript
const tuple: [number, number, number] = [1, 2, 3]
const namedTuple: [a: number, b: number, c: number] = [1, 2, 3]
```

The names are mostly useful for callers to differentiate between tuple values.

> What can I do to be helpful here? Are there some .ts or other files I should include in each package to make this process easier here?

You could write the type definitions directly in the package(s), either by providing an `index.d.ts` file in the package root that provides type definitions, or actually writing the source code in TypeScript!

This is the nice thing about DefinitelyTyped, though; it allows community contributions without forcing the package author/maintainer to adopt TypeScript.

**Comment by Myndex** [comment]
> > _The arrays passed to these functions may have additional elements all of which are ignored - only elements 0 1 2 are used._
> _~I actually avoided changing the type signature for the `[…]toY()` functions from `number[]` to tuples because of this!~
 On further reflection I decided to actually adopt the tuple types since it does make the function usage more obvious!_

As I'm less familiar with typescript — even typing them as tuples, the "bonus" discarded elements won't be an issue?

> > The " ...toY" functions all ignore any alpha, as alpha processing is separate in alphaBlend() where the first array is rgba and the second is rgb only.
> 
> I didn’t realise this, I can update the signature/docs to reflect this nuance

It was a recent decision... at one point I was going to handle the alpha processing inside the ...toY functions, but I am trying to adopt some more modern patterns now that this stuff is becoming public, such as more functions that do more granular work. And the alpha processing is something that is color space agnostic (so long as both colors are in the same space).


> > _calcAPCA() requires colorparsley […]_
> 
> _I didn’t include this function in the type definitions but can add it in the next revision. I think since colorParsley is used in the implementation but not in the type definition it shouldn’t be a problem, but we can document the requirement in the type documentation._




> _named tuples are still just comma-separated arrays, so these are equivalent:_
> 
> ```ts
> const tuple: [number, number, number] = [1, 2, 3]
> const namedTuple: [a: number, b: number, c: number] = [1, 2, 3]
> ```

Really? That's helpful. Guess I need to climb the TS learning curve a it more...


> _You could write the type definitions directly in the package(s), either by providing an `index.d.ts` file in the package root that provides type definitions, or actually writing the source code in TypeScript!_

Okay, I'd like to do this for colorparsley and also bridge-pca as well... Does in go in a special place?


> This is the nice thing about DefinitelyTyped, though; it allows community contributions without forcing the package author/maintainer to adopt TypeScript.

Now that some of this is stabilizing, I'm not opposed to doing some ports, and typescript seems like a good choice...



Thank you for all your help and input!!

Andu

**Review by peterblazejewicz — APPROVED** [review]
LGTM!

**Comment by daneden** [comment]
Ready to merge

**Pull request merged**